src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt

```diff
@@ -73,53 +83,131 @@ class VisitService(
   }
 
   fun cancelVisit(offenderNo: String, visitId: Long, visitDto: CancelVisitRequest) {
+    val today = LocalDate.now()
+    val visit = visitRepository.findById(visitId).orElseThrow(PrisonerNotFoundException(offenderNo))
+    val visitOutcome = visitOutcomeRepository.findById(VisitOutcomeReason.pk(visitDto.outcome)).orElseThrow()
+
+    val cancVisitStatus = visitStatusRepository.findById(VisitStatus.CANCELLED).orElseThrow()
+    val absEventOutcome = eventOutcomeRepository.findById(EventOutcome.ABS).orElseThrow()
+    val cancEventStatus = eventStatusRepository.findById(EventStatus.CANCELLED).orElseThrow()
+
+    if (visit.visitStatus?.code == "CANC") {
+      throw DataNotFoundException("Visit already cancelled") // or 2xx already done?
+    } else if (visit.visitStatus?.code != "SCH") {
+      throw DataNotFoundException("Visit already completed")
+    }
+
+    visit.visitStatus = cancVisitStatus
+
+    visit.visitors.forEach {
+      it.eventOutcome = absEventOutcome
+      it.eventStatus = cancEventStatus
+      it.outcomeReason = visitOutcome
+    }
+
+    val visitOrder = visit.visitOrder
+    if (visitOrder != null) {
+      visitOrder.status = cancVisitStatus
+      visitOrder.outcomeReason = visitOutcome
+      visitOrder.expiryDate = today
+
+      cancelBalance(visit, today)
+    }
   }
 
   private fun createBalance(
+    visit: Visit,
     visitDto: CreateVisitRequest,
     offenderBooking: OffenderBooking,
   ) {
-    val offenderVisitBalance = offenderVisitBalanceRepository.findById(offenderBooking.bookingId).orElseThrow()
+    val offenderVisitBalance = offenderVisitBalanceRepository.findById(offenderBooking.bookingId)
```

Comment:
very minor style thing; slightly better to use the Kotlin findByIdOrNull and not use the Java optional thing like `isPresent`

then the below just becomes

`
if (offenderVisitBalance != null) {
  if (offenderVisitBalance.remainingPrivilegedVisitOrders > 0) {
  }
}
`

Reply:
👍